File: ruby/tests/integration/test_integration.rb
$LOAD_PATH.unshift(File.expand_path('../..', __dir__))
require 'lira'
require 'minitest/autorun'

include Lira

PROJECT_ROOT = File.expand_path('../../..', __dir__)
REFERENCE = File.join(PROJECT_ROOT, 'tests', 'integration', 'reference.yaml')
CANONICALIZE = File.join(PROJECT_ROOT, 'tools', 'yaml_canonicalize.py')
OUTPUT = File.join(__dir__, 'integration.yaml')

class TestIntegration < Minitest::Test
  def test_roundtrip_from_reference
    ref_arch = ArchSerYaml.read_arch(REFERENCE)

    raw = "#{OUTPUT}.raw"
    begin
      ArchSerYaml.write_arch(ref_arch, raw)
      system('python3', CANONICALIZE, raw, OUTPUT)
      File.delete(raw)

      arch2 = ArchSerYaml.read_arch(OUTPUT)
      assert_equal ref_arch, arch2, 'round-trip equality failed'
    ensure
      File.delete(raw) if File.exist?(raw)
    end
  end
end

File: python/tests/integration/test_integration.py
import subprocess, sys
from pathlib import Path

from python.lira import arch_ser_yaml

PROJECT_ROOT = Path(__file__).parent.parent.parent.parent
REFERENCE = PROJECT_ROOT / "tests" / "integration" / "reference.yaml"
CANONICALIZE = PROJECT_ROOT / "tools" / "yaml_canonicalize.py"


def test_integration():
    ref_arch = arch_ser_yaml.read_arch(REFERENCE)

    output = PROJECT_ROOT / "python" / "tests" / "integration" / "integration.yaml"
    raw = output.with_suffix(".raw.yaml")
    try:
        arch_ser_yaml.write_arch(ref_arch, raw)
        subprocess.run(
            [sys.executable, str(CANONICALIZE), str(raw), str(output)], check=True
        )
        raw.unlink()

        arch2 = arch_ser_yaml.read_arch(output)
        assert ref_arch == arch2, "Round-trip failed"
    finally:
        raw.unlink(missing_ok=True)
